### DISCUSSION TITLE
Playback for live playing?

### DISCUSSION THREAD

**Discussion by brkl** [melanchall/drywetmidi#268] [Q&A] [answered]
Hello, and first of all thank you for your wonderful work on this project.
I have an application for generative music for my personal use. It's hacked together piece of software I've been writing on and off for years. I used an out of date and quite minimal MIDI library in the past and was able to switch to your library in an afternoon. It works just as well, but I'm not really making use of your features and I would love to. The problem is that DryWetMidi seems to largely be about playing back files or patterns that are prepared beforehand, not on the fly. 

So right now I'm just sending note on events like this:
`NoteOnEvent thisEvent = new NoteOnEvent(pitch.NoteNumber, (SevenBitNumber)vol).Channel;
thisEvent.Channel = (FourBitNumber)channel;
midiOut.SendEvent(thisEvent);`
((Is there a better way to change the midi channel?))

Instead of that, I would like to make use of your classes such as [Note](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Interaction.Note.html) and pass that to a Playback or some other object that would send the Note On message immediately and deal with the Note Off message when the time came. I'm doing that right now with my own classes, but almost certainly more poorly than your library does. I'd also like to use Playback's timers to keep Tempo and send events at 16th or 64th note intervals so my system would keep time better than it does now. Of course this would allow me to use your Music Theory classes and other tools which seem very useful for me.

So I suppose what I'm asking is a different kind of Pattern that I do not prepare before hand with notes but add notes (and other events) as a Playback is playing the Pattern. 

Does this seems feasible on DryWetMidi? Is there already a way to do something like this that I'm not seeing?

Thanks again for your work on this library.

Regards,
brkl

**Comment by melanchall** [comment, marked as answer]
Hi,

> Is there a better way to change the midi channel?

No, this is how a channel should be set. You can of course use object initialization of C#:

```csharp
midiOut.SendEvent(new NoteOnEvent(pitch.NoteNumber, (SevenBitNumber)vol)
{
    Channel = (FourBitNumber)channel
});
```

> Does this seems feasible on DryWetMidi? Is there already a way to do something like this that I'm not seeing?

Unfortunately there is no way to add objects into `Playback` on the fly. There is [the issue](https://github.com/melanchall/drywetmidi/issues/78) about that, but I have no plans currently to work on it.

As for sending a single `Note`, it's possible:

```csharp
var playback = new[] { note }.GetPlayback(...);
playback.Start();
```

Thanks,
Max

**Reply by brkl** [reply]
Thank you for your answer! I suppose I will work on other things and hope for this feature to be implemented in the future. 

Looking at Playback's OnClockTicked, it seems to read through a list of events, sending them off when the time is right and then stop playback or start looping depending on that setting. I was thinking of reading an entire list of events each time the clock is ticked and removing each event from the list when they are sent off and not stopping if the list is empty. Of course that would require a way to add events to the list and would get slow if the list was populated with many events. 

Anyway, I'm sure you have your list of priorities, developing a library is beyond me. Thank you for the coding tip regarding initialization, I will make use of it.

Regards,
brkl
